src/template/transaction/executor.ts

```diff
@@ -786,5 +746,228 @@ export async function execSetVarInstruction(
+
+  await selectTokens(interpreter, ctx, amount, options, action.autoChange, changeAddress);
+
+  return actual;
+}
+
+async function validateWithdrawalNanoAction(
+  interpreter: ITxTemplateInterpreter,
+  ctx: TxTemplateContext,
+  action: z.infer<typeof NanoWithdrawalAction>
+) {
+  const token = getVariable<string>(action.token, ctx.vars, NanoWithdrawalAction.shape.token);
+  const amount = getVariable<OutputValueType>(
+    action.amount,
+    ctx.vars,
+    NanoWithdrawalAction.shape.amount
+  );
+  const address =
+    getVariable<string | undefined>(action.address, ctx.vars, NanoWithdrawalAction.shape.address) ??
+    (await interpreter.getAddress());
+
+  // This is the action without variables, which will be used to create the header
+  const actual = {
+    ...action,
+    token,
+    amount,
+    address,
+  };
+
+  const tokenData = action.useCreatedToken ? 1 : ctx.addToken(token);
+  const script = createOutputScriptFromAddress(address, interpreter.getNetwork());
+  const output = new Output(amount, script, { tokenData });
+  ctx.addOutputs(-1, output);
+  return actual;
+}
+
+async function validateGrantAuthorityNanoAction(
+  interpreter: ITxTemplateInterpreter,
+  ctx: TxTemplateContext,
+  action: z.infer<typeof NanoGrantAuthorityAction>
+) {
+  const token = getVariable<string>(action.token, ctx.vars, NanoGrantAuthorityAction.shape.token);
+  const { authority } = action;
+  const address = getVariable<string | undefined>(
+    action.address,
+    ctx.vars,
+    NanoGrantAuthorityAction.shape.address
+  );
+
+  // This is the action without variables, which will be used to create the header
+  const actual = {
+    ...action,
+    token,
+    authority,
+    address,
+  };
+
+  if (action.skipSelection) {
+    // Do not select inputs
+    return actual;
+  }
+
+  let authoritiesInt = 0n;
+  if (authority === 'mint') {
+    authoritiesInt += TOKEN_MINT_MASK;
+  }
+  if (authority === 'melt') {
+    authoritiesInt += TOKEN_MELT_MASK;
+  }
+
+  // Find utxos
+  const options: IGetUtxosOptions = {
+    token,
+    authorities: authoritiesInt,
+  };
+  if (address) {
+    options.filter_address = address;
+  }
+  await selectAuthorities(interpreter, ctx, options);
+  return actual;
+}
+
+async function validateAcquireAuthorityNanoAction(
+  interpreter: ITxTemplateInterpreter,
+  ctx: TxTemplateContext,
+  action: z.infer<typeof NanoAcquireAuthorityAction>
+) {
+  const token = getVariable<string>(action.token, ctx.vars, NanoAcquireAuthorityAction.shape.token);
+  const address =
+    getVariable<string | undefined>(
+      action.address,
+      ctx.vars,
+      NanoAcquireAuthorityAction.shape.address
+    ) ?? (await interpreter.getAddress());
+
+  // This is the action without variables, which will be used to create the header
+  const actual = {
+    ...action,
+    token,
+    authority: action.authority,
+    address,
+  };
+
+  const tokenData = TOKEN_AUTHORITY_MASK | (action.useCreatedToken ? 1 : ctx.addToken(token));
```

Comment:
I'm thinking about this, I don't think we can use the created token here. This is to acquire an authority from the contract, so the token must already have been created. Does it make sense?

Reply:
Makes sense but this works only for acquire and withdrawal, correct?

Reply:
Yes!